File: web/src/types/sale.ts
// Sale types - mirrors internal/domain/sale.go

import type { Branch } from './branch'
import type { Customer } from './customer'
import type { Item } from './item'
import type { PaymentMethod } from './payment'

export type SaleStatus = 'completed' | 'pending' | 'cancelled' | 'refunded' | 'partial_refund'
export type SaleType = 'direct' | 'layaway'

export const SALE_STATUSES: { value: SaleStatus; label: string; color: string }[] = [
  { value: 'completed', label: 'Completada', color: 'green' },
  { value: 'pending', label: 'Pendiente', color: 'yellow' },
  { value: 'cancelled', label: 'Cancelada', color: 'gray' },
  { value: 'refunded', label: 'Reembolsada', color: 'red' },
  { value: 'partial_refund', label: 'Reembolso Parcial', color: 'orange' },
]

export const SALE_TYPES: { value: SaleType; label: string }[] = [
  { value: 'direct', label: 'Venta Directa' },
  { value: 'layaway', label: 'Apartado' },
]

export interface Sale {
  id: number
  branch_id: number
  item_id: number
  customer_id?: number
  sale_number: string
  sale_type: SaleType

  // Pricing
  sale_price: number
  discount_amount: number
  discount_reason?: string
  final_price: number

  // Payment
  payment_method: PaymentMethod
  reference_number?: string

  // Status
  status: SaleStatus
  sale_date: string

  // Refund info
  refund_amount?: number
  refund_reason?: string
  refunded_at?: string
  refunded_by?: number

  // Notes
  notes?: string

  // Cash session reference
  cash_session_id?: number

  // Audit
  created_by?: number
  updated_by?: number
  created_at: string
  updated_at: string
  deleted_at?: string

  // Relations
  branch?: Branch
  item?: Item
  customer?: Customer
}

export interface CreateSaleInput {
  branch_id: number
  item_id: number
  customer_id?: number
  sale_type?: SaleType
  sale_price?: number
  discount_amount?: number
  discount_reason?: string
  payment_method: PaymentMethod
  reference_number?: string
  notes?: string
}

export interface RefundSaleInput {
  amount?: number
  reason: string
}

export interface SaleListParams {
  page?: number
  per_page?: number
  order_by?: string
  order?: 'asc' | 'desc'
  search?: string
  branch_id?: number
  item_id?: number
  customer_id?: number
  status?: SaleStatus
  sale_type?: SaleType
  payment_method?: PaymentMethod
  date_from?: string
  date_to?: string
}

export interface SalesSummary {
  total_sales: number
  total_amount: number
  total_discount: number
  net_amount: number
  total_refunds: number
  refund_amount: number
  sales_by_method: {
    method: PaymentMethod
    count: number
    amount: number
  }[]
}

File: internal/domain/sale.go
package domain

import (
	"time"
)

// SaleStatus represents the status of a sale
type SaleStatus string

const (
	SaleStatusCompleted     SaleStatus = "completed"
	SaleStatusPending       SaleStatus = "pending"
	SaleStatusCancelled     SaleStatus = "cancelled"
	SaleStatusRefunded      SaleStatus = "refunded"
	SaleStatusPartialRefund SaleStatus = "partial_refund"
)

// Sale represents a sale of an item
type Sale struct {
	ID         int64  `json:"id"`
	BranchID   int64  `json:"branch_id"`
	ItemID     int64  `json:"item_id"`
	CustomerID *int64 `json:"customer_id,omitempty"` // Optional customer
	SaleNumber string `json:"sale_number"`
	SaleType   string `json:"sale_type"` // direct, layaway

	// Pricing
	SalePrice       float64  `json:"sale_price"`
	DiscountAmount  float64  `json:"discount_amount"`
	DiscountReason  *string  `json:"discount_reason,omitempty"`
	FinalPrice      float64  `json:"final_price"`

	// Payment
	PaymentMethod   PaymentMethod `json:"payment_method"`
	ReferenceNumber *string       `json:"reference_number,omitempty"`

	// Status
	Status   SaleStatus `json:"status"`
	SaleDate time.Time  `json:"sale_date"`

	// Refund info
	RefundAmount  *float64   `json:"refund_amount,omitempty"`
	RefundReason  *string    `json:"refund_reason,omitempty"`
	RefundedAt    *time.Time `json:"refunded_at,omitempty"`
	RefundedBy    *int64     `json:"refunded_by,omitempty"`

	// Notes
	Notes *string `json:"notes,omitempty"`

	// Cash session reference
	CashSessionID *int64 `json:"cash_session_id,omitempty"`

	// Audit
	CreatedBy int64 `json:"created_by,omitempty"`
	UpdatedBy int64 `json:"updated_by,omitempty"`

	// Timestamps
	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
	DeletedAt *time.Time `json:"deleted_at,omitempty"`

	// Relations
	Branch   *Branch   `json:"branch,omitempty"`
	Item     *Item     `json:"item,omitempty"`
	Customer *Customer `json:"customer,omitempty"`
}

// TableName returns the database table name
func (Sale) TableName() string {
	return "sales"
}

// IsRefunded checks if the sale has been refunded
func (s *Sale) IsRefunded() bool {
	return s.Status == SaleStatusRefunded
}

// CanBeRefunded checks if the sale can be refunded
func (s *Sale) CanBeRefunded() bool {
	return s.Status == SaleStatusCompleted
}
